modules/sor/sor.service.ts

```diff
@@ -1,96 +1,178 @@
+    private async getBestSwapPathFromBothVersions(input: Omit<GetSwapPathsInput, 'protocolVersion'>): Promise<{
+        paths: PathWithAmount[] | null;
+        protocolVersion: number;
+    }> {
+        const pathsV2 = await this.getSwapPathsWithRetry({ ...input, protocolVersion: 2 });
+        const pathsV3 = await this.getSwapPathsWithRetry({ ...input, protocolVersion: 3 });
+
         if (input.swapType === 'EXACT_IN') {
-            return parseFloat(swapBalancerV2.returnAmount) > parseFloat(swapBalancerV3.returnAmount)
-                ? swapBalancerV2
-                : swapBalancerV3;
+            return parseFloat(pathsV2.returnAmount) > parseFloat(pathsV3.returnAmount) ? pathsV2 : pathsV3;
         } else {
             // return swap path with smallest non-zero amountsIn (if it exists)
-            if (parseFloat(swapBalancerV2.returnAmount) === 0) {
-                return swapBalancerV3;
-            } else if (parseFloat(swapBalancerV3.returnAmount) === 0) {
-                return swapBalancerV2;
+            if (parseFloat(pathsV2.returnAmount) === 0) {
+                return pathsV3;
+            } else if (parseFloat(pathsV3.returnAmount) === 0) {
+                return pathsV2;
             } else {
-                return parseFloat(swapBalancerV2.returnAmount) < parseFloat(swapBalancerV3.returnAmount)
-                    ? swapBalancerV2
-                    : swapBalancerV3;
+                return parseFloat(pathsV3.returnAmount) < parseFloat(pathsV3.returnAmount) ? pathsV2 : pathsV3;
+            }
+        }
+    }
+
+    private async getSwapPathsWithRetry(
+        input: GetSwapPathsInput,
+    ): Promise<{ paths: PathWithAmount[] | null; protocolVersion: number; returnAmount: string }> {
+        try {
+            const { pools: poolsFromDb, underlyingTokens } = await getBasePoolsFromDb(
+                input.chain,
+                input.protocolVersion,
+                input.considerPoolsWithHooks,
+                input.poolIds,
+            );
+
+            const tokenIn = await getToken(input.tokenIn as Address, input.chain);
+            const tokenOut = await getToken(input.tokenOut as Address, input.chain);
+            const swapKind = mapSwapKind(input.swapType);
+
+            // retry with different max depth if no paths are found
+            let swapOptions = this.buildSwapOptions(input.graphTraversalConfig);
+            let paths = await SOR.getPathsWithPools(
+                tokenIn,
+                tokenOut,
+                swapKind,
+                input.swapAmount.amount,
+                poolsFromDb,
+                underlyingTokens,
+                input.protocolVersion,
+                swapOptions,
+            );
+
+            if (!paths) {
+                swapOptions = this.buildSwapOptions({
+                    ...swapOptions.graphTraversalConfig,
+                    maxNonBoostedPathDepth: swapOptions.graphTraversalConfig.maxNonBoostedPathDepth! + 1,
```

Comment:
the graphTraversalConfig is never set and defaults are always used. We should extract this into the constants or config. We dont want it to be adjustable in the query as it can basically kill the SOR.